File: src/services/arService.ts
/**
 * AR Service
 * API calls for AR visualization, furniture placement, and interaction
 */

import { post, get } from './api';

export interface ARFurniture {
  id: string;
  name: string;
  modelUrl: string;
  scale: { x: number; y: number; z: number };
  position: { x: number; y: number; z: number };
  rotation: { x: number; y: number; z: number };
}

export interface ARScene {
  id: string;
  designId: string;
  furnitureList: ARFurniture[];
  roomDimensions: {
    width: number;
    height: number;
    length: number;
  };
  lightingSetup: {
    ambientLight: number;
    spotlights: Array<{
      position: { x: number; y: number; z: number };
      intensity: number;
    }>;
  };
  textures: {
    floorMaterial: string;
    wallColor: string;
  };
  createdAt: string;
  updatedAt: string;
}

/**
 * Get AR models for furniture
 */
export async function getARModels(designId: string) {
  return get<ARFurniture[]>(`/api/ar/models?designId=${designId}`);
}

/**
 * Get specific AR model
 */
export async function getARModel(modelId: string) {
  return get<ARFurniture>(`/api/ar/models/${modelId}`);
}

/**
 * Create AR scene from design
 */
export async function createARScene(designId: string) {
  return post<ARScene>('/api/ar/scenes', { designId });
}

/**
 * Get AR scene
 */
export async function getARScene(sceneId: string) {
  return get<ARScene>(`/api/ar/scenes/${sceneId}`);
}

/**
 * Update furniture position in scene
 */
export async function updateFurniturePosition(
  sceneId: string,
  furnitureId: string,
  position: { x: number; y: number; z: number }
) {
  return post(`/api/ar/scenes/${sceneId}/furniture/${furnitureId}/position`, {
    position,
  });
}

/**
 * Rotate furniture in scene
 */
export async function rotateFurniture(
  sceneId: string,
  furnitureId: string,
  rotation: { x: number; y: number; z: number }
) {
  return post(`/api/ar/scenes/${sceneId}/furniture/${furnitureId}/rotation`, {
    rotation,
  });
}

/**
 * Scale furniture in scene
 */
export async function scaleFurniture(
  sceneId: string,
  furnitureId: string,
  scale: { x: number; y: number; z: number }
) {
  return post(`/api/ar/scenes/${sceneId}/furniture/${furnitureId}/scale`, {
    scale,
  });
}

/**
 * Add furniture to scene
 */
export async function addFurnitureToScene(
  sceneId: string,
  furnitureId: string,
  position?: { x: number; y: number; z: number }
) {
  return post<ARFurniture>(
    `/api/ar/scenes/${sceneId}/furniture`,
    { furnitureId, position }
  );
}

/**
 * Remove furniture from scene
 */
export async function removeFurnitureFromScene(
  sceneId: string,
  furnitureId: string
) {
  return post(`/api/ar/scenes/${sceneId}/furniture/${furnitureId}/remove`);
}

/**
 * Get furniture collision detection
 */
export async function checkCollisions(
  sceneId: string,
  furnitureId: string,
  position: { x: number; y: number; z: number }
) {
  return post<{
    hasCollision: boolean;
    collidingWith?: string[];
  }>(`/api/ar/scenes/${sceneId}/furniture/${furnitureId}/collision-check`, {
    position,
  });
}

/**
 * Generate optimal layout
 */
export async function generateOptimalLayout(sceneId: string) {
  return post<{ furniturePositions: Record<string, any> }>(
    `/api/ar/scenes/${sceneId}/optimize-layout`
  );
}

/**
 * Capture AR screenshot
 */
export async function captureARScreenshot(sceneId: string, imageData: Blob) {
  const formData = new FormData();
  formData.append('image', imageData);
  return post<{ url: string }>(
    `/api/ar/scenes/${sceneId}/screenshot`,
    formData
  );
}

/**
 * Export AR scene as 3D model
 */
export async function exportARScene(
  sceneId: string,
  format: 'gltf' | 'usdz' | 'obj'
) {
  return get<Blob>(`/api/ar/scenes/${sceneId}/export?format=${format}`);
}

/**
 * Get supported device capabilities
 */
export async function getDeviceCapabilities() {
  return post<{
    supportsAR: boolean;
    supportsWebGL: boolean;
    maxTextureSize: number;
    gpuInfo?: string;
  }>('/api/ar/device-capabilities', navigator.userAgent);
}

/**
 * Check AR support
 */
export async function checkARSupport() {
  const capabilities = await getDeviceCapabilities();
  return capabilities.success && capabilities.data?.supportsAR;
}

export default {
  getARModels,
  getARModel,
  createARScene,
  getARScene,
  updateFurniturePosition,
  rotateFurniture,
  scaleFurniture,
  addFurnitureToScene,
  removeFurnitureFromScene,
  checkCollisions,
  generateOptimalLayout,
  captureARScreenshot,
  exportARScene,
  getDeviceCapabilities,
  checkARSupport,
};

File: backend/services/ar_service.py
"""
AR Visualization Service - Generate AR models and scenes
"""

import json
import logging
from typing import Dict, List, Any, Optional
from datetime import datetime

logger = logging.getLogger(__name__)

class ARVisualizationService:
    """Handle AR scene creation and furniture placement"""
    
    def __init__(self):
        self.logger = logger
    
    def create_ar_scene(
        self,
        room_dimensions: Dict[str, float],
        furniture_list: List[Dict[str, Any]],
        design_colors: Dict[str, str]
    ) -> Dict[str, Any]:
        """
        Create AR scene with room dimensions and furniture
        Returns 3D model data (GLTF/USDZ format)
        """
        
        try:
            scene = {
                'scene_id': self._generate_scene_id(),
                'type': 'interior_design',
                'version': '1.0',
                'room': {
                    'dimensions': room_dimensions,
                    'walls': self._create_walls(room_dimensions, design_colors),
                    'floor': self._create_floor(room_dimensions, design_colors),
                    'ceiling': self._create_ceiling(room_dimensions, design_colors)
                },
                'objects': self._create_furniture_models(furniture_list),
                'lighting': self._create_lighting_setup(room_dimensions),
                'camera': {
                    'position': [0, 1.6, 0],
                    'lookAt': [0, 0.8, 0],
                    'fov': 60
                },
                'metadata': {
                    'created_at': datetime.now().isoformat(),
                    'format': 'gltf'
                }
            }
            
            return {
                'success': True,
                'scene': scene,
                'scene_id': scene['scene_id'],
                'format': 'gltf/usdz'
            }
        
        except Exception as e:
            self.logger.error(f'AR scene creation error: {e}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def update_furniture_position(
        self,
        scene_id: str,
        furniture_id: str,
        position: Dict[str, float]
    ) -> Dict[str, Any]:
        """Update furniture position in AR scene"""
        
        try:
            # Validate position
            if not all(k in position for k in ['x', 'y', 'z']):
                raise ValueError('Position must have x, y, z coordinates')
            
            return {
                'success': True,
                'scene_id': scene_id,
                'furniture_id': furniture_id,
                'position': position,
                'timestamp': datetime.now().isoformat()
            }
        
        except Exception as e:
            self.logger.error(f'Position update error: {e}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def detect_placement_collision(
        self,
        furniture: Dict[str, Any],
        position: Dict[str, float],
        room_dimensions: Dict[str, float]
    ) -> Dict[str, Any]:
        """
        Detect if furniture placement causes collision
        Uses simple AABB collision detection
        """
        
        try:
            # Get furniture bounds
            furniture_width = furniture.get('width', 1.0)
            furniture_depth = furniture.get('depth', 1.0)
            furniture_height = furniture.get('height', 1.0)
            
            # Check room bounds
            room_x = room_dimensions.get('width', 5.0)
            room_y = room_dimensions.get('height', 2.8)
            room_z = room_dimensions.get('length', 5.0)
            
            x, y, z = position.get('x', 0), position.get('y', 0), position.get('z', 0)
            
            # Check if furniture fits in room
            has_collision = False
            collision_details = []
            
            if x + furniture_width / 2 > room_x / 2 or x - furniture_width / 2 < -room_x / 2:
                has_collision = True
                collision_details.append('Furniture exceeds room width')
            
            if z + furniture_depth / 2 > room_z / 2 or z - furniture_depth / 2 < -room_z / 2:
                has_collision = True
                collision_details.append('Furniture exceeds room depth')
            
            if y + furniture_height > room_y:
                has_collision = True
                collision_details.append('Furniture exceeds room height')
            
            return {
                'success': True,
                'has_collision': has_collision,
                'collision_details': collision_details,
                'furniture_id': furniture.get('id', 'unknown')
            }
        
        except Exception as e:
            self.logger.error(f'Collision detection error: {e}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def generate_design_screenshot(
        self,
        scene: Dict[str, Any],
        camera_angle: Dict[str, float]
    ) -> Dict[str, Any]:
        """Generate 2D screenshot from 3D scene"""
        
        try:
            # In real implementation, use Three.js or similar to render
            # For now, return metadata for screenshot
            
            return {
                'success': True,
                'screenshot_url': f'/api/ar/scenes/{scene.get("scene_id", "unknown")}/render.jpg',
                'resolution': '1920x1080',
                'camera_angle': camera_angle,
                'format': 'jpeg'
            }
        
        except Exception as e:
            self.logger.error(f'Screenshot generation error: {e}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def export_scene(
        self,
        scene: Dict[str, Any],
        export_format: str = 'gltf'
    ) -> Dict[str, Any]:
        """
        Export AR scene in different formats
        Supports: gltf, usdz (iOS), glb
        """
        
        try:
            if export_format not in ['gltf', 'usdz', 'glb']:
                raise ValueError(f'Unsupported format: {export_format}')
            
            scene_id = scene.get('scene_id', 'unknown')
            
            return {
                'success': True,
                'download_url': f'/api/ar/scenes/{scene_id}/export/{export_format}',
                'format': export_format,
                'file_format': {
                    'gltf': 'text/plain',  # GLTF is text-based
                    'glb': 'application/octet-stream',
                    'usdz': 'model/vnd.usdz+zip'
                }.get(export_format),
                'ready': True
            }
        
        except Exception as e:
            self.logger.error(f'Scene export error: {e}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def _create_walls(self, dimensions: Dict[str, float], colors: Dict[str, str]) -> List[Dict[str, Any]]:
        """Create wall models for AR scene"""
        width = dimensions.get('width', 5.0)
        height = dimensions.get('height', 2.8)
        length = dimensions.get('length', 5.0)
        
        wall_color = colors.get('walls', '#F5F5F5')
        
        walls = []
        
        # Front and back walls
        walls.append({
            'id': 'wall_front',
            'type': 'plane',
            'position': [0, height / 2, length / 2],
            'rotation': [0, 0, 0],
            'scale': [width, height, 0.1],
            'color': wall_color,
            'material': 'standard'
        })
        
        walls.append({
            'id': 'wall_back',
            'type': 'plane',
            'position': [0, height / 2, -length / 2],
            'rotation': [0, 3.14159, 0],
            'scale': [width, height, 0.1],
            'color': wall_color,
            'material': 'standard'
        })
        
        # Left and right walls
        walls.append({
            'id': 'wall_left',
            'type': 'plane',
            'position': [-width / 2, height / 2, 0],
            'rotation': [0, 1.5708, 0],
            'scale': [length, height, 0.1],
            'color': wall_color,
            'material': 'standard'
        })
        
        walls.append({
            'id': 'wall_right',
            'type': 'plane',
            'position': [width / 2, height / 2, 0],
            'rotation': [0, -1.5708, 0],
            'scale': [length, height, 0.1],
            'color': wall_color,
            'material': 'standard'
        })
        
        return walls
    
    def _create_floor(self, dimensions: Dict[str, float], colors: Dict[str, str]) -> Dict[str, Any]:
        """Create floor model for AR scene"""
        width = dimensions.get('width', 5.0)
        length = dimensions.get('length', 5.0)
        floor_color = colors.get('floor', '#8B7355')
        
        return {
            'id': 'floor',
            'type': 'plane',
            'position': [0, 0, 0],
            'rotation': [0, 0, 0],
            'scale': [width, 0.1, length],
            'color': floor_color,
            'material': 'floor'
        }
    
    def _create_ceiling(self, dimensions: Dict[str, float], colors: Dict[str, str]) -> Dict[str, Any]:
        """Create ceiling model for AR scene"""
        width = dimensions.get('width', 5.0)
        height = dimensions.get('height', 2.8)
        length = dimensions.get('length', 5.0)
        ceiling_color = colors.get('ceiling', '#FFFFFF')
        
        return {
            'id': 'ceiling',
            'type': 'plane',
            'position': [0, height, 0],
            'rotation': [0, 0, 0],
            'scale': [width, 0.1, length],
            'color': ceiling_color,
            'material': 'ceiling'
        }
    
    def _create_furniture_models(self, furniture_list: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Create furniture models for AR scene"""
        models = []
        
        furniture_defaults = {
            'sofa': {'width': 2.5, 'depth': 1.0, 'height': 0.8, 'model': 'sofa.gltf'},
            'table': {'width': 1.2, 'depth': 1.2, 'height': 0.75, 'model': 'table.gltf'},
            'chair': {'width': 0.6, 'depth': 0.6, 'height': 0.9, 'model': 'chair.gltf'},
            'shelf': {'width': 1.0, 'depth': 0.4, 'height': 2.0, 'model': 'shelf.gltf'},
            'lamp': {'width': 0.3, 'depth': 0.3, 'height': 1.5, 'model': 'lamp.gltf'},
            'bed': {'width': 1.4, 'depth': 2.0, 'height': 0.5, 'model': 'bed.gltf'},
        }
        
        for idx, item in enumerate(furniture_list):
            furniture_type = item.get('type', 'chair').lower()
            defaults = furniture_defaults.get(furniture_type, furniture_defaults['chair'])
            
            model = {
                'id': item.get('id', f'furniture_{idx}'),
                'name': item.get('name', furniture_type),
                'type': furniture_type,
                'position': item.get('position', [0, defaults['height'] / 2, idx]),
                'rotation': item.get('rotation', [0, 0, 0]),
                'scale': [
                    item.get('width', defaults['width']),
                    item.get('height', defaults['height']),
                    item.get('depth', defaults['depth'])
                ],
                'color': item.get('color', '#FFFFFF'),
                'material': item.get('material', 'standard'),
                'model_url': f'/models/{defaults["model"]}',
                'interactive': True
            }
            
            models.append(model)
        
        return models
    
    def _create_lighting_setup(self, dimensions: Dict[str, float]) -> List[Dict[str, Any]]:
        """Create lighting setup for AR scene"""
        height = dimensions.get('height', 2.8)
        width = dimensions.get('width', 5.0)
        length = dimensions.get('length', 5.0)
        
        lights = [
            {
                'id': 'ambient',
                'type': 'ambient',
                'intensity': 0.8,
                'color': '#FFFFFF'
            },
            {
                'id': 'main_light',
                'type': 'directional',
                'position': [0, height, length / 2],
                'target': [0, 0, 0],
                'intensity': 1.0,
                'color': '#FFFFFF',
                'shadow': True
            },
            {
                'id': 'accent_light_1',
                'type': 'point',
                'position': [width / 2, height - 0.5, 0],
                'intensity': 0.5,
                'color': '#FFFACD',
                'range': 10
            },
            {
                'id': 'accent_light_2',
                'type': 'point',
                'position': [-width / 2, height - 0.5, 0],
                'intensity': 0.5,
                'color': '#FFFACD',
                'range': 10
            }
        ]
        
        return lights
    
    def _generate_scene_id(self) -> str:
        """Generate unique scene ID"""
        import uuid
        return f"scene_{uuid.uuid4().hex[:8]}"


# Initialize AR service
ar_service = ARVisualizationService()
